packages/neovim/src/attach/attach.test.ts

```diff
@@ -154,19 +154,27 @@ describe('Nvim API', () => {
     expect(newLines).toEqual(['line1', 'line2']);
   });
 
-  it('emits "disconnect" after quit', done => {
+  it('emits "disconnect" after quit', async () => {
     const disconnectMock = jest.fn();
     nvim.on('disconnect', disconnectMock);
     nvim.quit();
 
-    proc.on('close', () => {
-      expect(disconnectMock.mock.calls.length).toBe(1);
-      done();
+    const closePromise = new Promise(resolve => {
+      proc.on('close', resolve);
     });
 
-    // Event doesn't actually emit when we quit nvim, but when the child process is killed
-    if (proc && proc.connected) {
-      proc.disconnect();
-    }
+    let timeout: NodeJS.Timeout;
+
+    await Promise.race([
```

Comment:
Thanks for looking at this. I think the `done` mechanism works, the problem is either in our test logic or there's a race during "teardown" so e.g. the EventEmitter event is not getting triggered.

Or maybe it's related to jest's "mocking" implementation, so removing the use of `jest.fn` could help (for most of our tests we have no reason to use it, especially this one).

Reply:
If we remove the mock function, how do we make sure `disconnect` is indeed emitted?

Reply:
just define a real function: 

```diff
diff --git a/packages/neovim/src/attach/attach.test.ts b/packages/neovim/src/attach/attach.test.ts
index 0ae7cc029623..1146b13e9263 100644
--- a/packages/neovim/src/attach/attach.test.ts
+++ b/packages/neovim/src/attach/attach.test.ts
@@ -183,12 +183,14 @@ describe('Nvim API', () => {
   });
 
   it('emits "disconnect" after quit', done => {
-    const disconnectMock = jest.fn();
-    nvim.on('disconnect', disconnectMock);
+    let calls = 0;
+    nvim.on('disconnect', function() {
+      calls = calls + 1;
+    });
     nvim.quit();
 
     proc.on('close', () => {
-      expect(disconnectMock.mock.calls.length).toBe(1);
+      expect(calls).toBe(1);
       done();
     });
 ```

however now that I look at it, it's obvious that the mock isn't the issue here, though I think it's an unnecessary layer to introduce to the test.